hashes/src/sha256.rs

```diff
@@ -74,41 +74,60 @@ impl HashEngine {
         }
     }
 
-    /// Create a new [`HashEngine`] from a [`Midstate`].
+    /// Creates a new [`HashEngine`] from a [`Midstate`].
     ///
-    /// # Panics
-    ///
-    /// If `length` is not a multiple of the block size.
-    pub fn from_midstate(midstate: Midstate, length: usize) -> HashEngine {
-        assert!(length % BLOCK_SIZE == 0, "length is no multiple of the block size");
-
+    /// This function should not be used directly unless you really know what you're doing i.e., you
+    /// probably should be a professional cryptographer.
```

Comment:
Sounds too gatekeepish to me. Having at least an example where this is useful (tags) would be nicer.

Reply:
```
/// The [`Midstate`] type is obscure and specialized and should not be used unless you are sure of what you are d
doing.
///
/// It represents "partially hashed data" but does not itself have properties of cryptographic hashes. For example,
/// when (ab)used as hashes, midstates are vulnerable to trivial length-extension attacks. They are typically used
/// to optimize the computation of full hashes. For example, when implementing BIP-340 tagged hashes, which
/// always begin by hashing the same fixed 64-byte prefix, it makes sense to hash the prefix once, store
/// the midstate as a constant, and hash any future data starting from the constant rather than from a fresh
/// hash engine.
```

Reply:
Yeah, that sounds good. Plus a note about not really needing to do that manually for BIP-340 since our `sha256t_newtype` macro does the hashing for you in `const` context.